File: LeetCode/Problems/Python/1095. Find in Mountain Array.py
#Binary search
#Time: O(logn)
#Space: O(1)


# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountain_arr: 'MountainArray') -> int:
        
        l = 0
        r = mountain_arr.length()-1
        #find i first, and find target in left, then right
        #if can't find in both, return -1
        
        while l < r:
            mid = l + (r-l)//2
            
            if mountain_arr.get(mid) > mountain_arr.get(mid+1):
                r = mid 
            else:
                l = mid + 1
        mount = l

        #find in left
        l = 0
        r = mount        
        
        while l <= r:
            mid = l + (r-l)//2
            if mountain_arr.get(mid) ==target:
                return mid
            elif mountain_arr.get(mid) < target:
                l = mid + 1
            else:
                r = mid - 1
                
                
        #find in right
        l = mount
        r = mountain_arr.length()-1  
        while l <= r:
            mid = l + (r-l)//2
            if mountain_arr.get(mid) ==target:
                return mid
            elif mountain_arr.get(mid) < target:
                r = mid - 1
            else:
                l = mid + 1
        
        return -1

File: LeetCode/Problems/C++/1095. Find in Mountain Array.cpp
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */

class Solution {
public:
    int findInMountainArray(int target, MountainArray &mountainArr) {
        int l = 0;
        int r = mountainArr.length()-1;
        int mid;
        
        //find peak
        while(l<r){
            mid = l + (r-l)/2;
            if(mountainArr.get(mid)>mountainArr.get(mid+1))
                r = mid;
            else
                l = mid+1;
        }
        
        int pivot = l;
        
        l = 0;
        r = pivot;
        
        //search in left
        while(l<=r){
            mid = l + (r-l)/2;
            if(mountainArr.get(mid) == target) 
                return mid;
            else if(mountainArr.get(mid) < target)
                l = mid+1;
            else
                r = mid-1;
        }
        
        //search in right
        l = pivot+1;
        r = mountainArr.length()-1;
        
        while(l<=r){
            mid = l + (r-l)/2;
            if(mountainArr.get(mid) == target) 
                return mid;
            else if(mountainArr.get(mid) < target)
                r = mid-1;
            else
                l = mid+1;
        }
        
        
        
        
        return -1;
    }
};